url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/layout/JGraphWordBrowser.java

```diff
@@ -0,0 +1,480 @@
+
+			// Includes all connections to/from the cell
+			Set cells = DefaultGraphModel.getEdges(this,
+					new Object[] { vertex });
+			cells.add(vertex);
+
+			// Removes the cells from the model
+			remove(cells.toArray());
+		}
+		return vertex;
+	}
+
+	/**
+	 * Returns the default port for the specified vertex.
+	 */
+	public Object getDefaultPort(Object vertex) {
+		return getChild(vertex, 0);
+	}
+
+	/**
+	 *
+	 * 
+	 */
+	public void add(String word, String wordType) {
+		if (word.endsWith(")")) {
+			int index = word.indexOf(" (");
+			wordType = word.substring(index + 1);
+			word = word.substring(0, index);
+		}
+		try {
+			List page = getContent(word);
+			Iterator it = page.iterator();
+			String type = "";
+			Object vertex = null;
+			while (it.hasNext()) {
+				String line = (String) it.next();
+				if (line.indexOf("<h3>Noun</h3>") >= 0) {
+					type = "(N)";
+					if (wordType.equals(type)) {
+						vertex = addVertex(word + " " + type);
+					} else {
+						vertex = null;
+					}
+				} else if (line.indexOf("<h3>Adjective</h3>") >= 0) {
+					type = "(A)";
+					if (wordType.equals(type)) {
+						vertex = addVertex(word + " " + type);
+					} else {
+						vertex = null;
+					}
+				} else if (line.indexOf("<h3>Verb</h3>") >= 0) {
+					type = "(V)";
+					if (wordType.equals(type)) {
+						vertex = addVertex(word + " " + type);
+					} else {
+						vertex = null;
+					}
+				}
+				if (vertex != null) {
+					if (line.indexOf("<li>") >= 0) {
+						List words = getWords(line);
+						String edge = "";
+						int index = line.lastIndexOf("(") + 1;
+						if (index >= 0) {
+							edge = line.substring(index, line.indexOf(")",
+									index));
+						}
+						int counter = 1;
+						Iterator it2 = words.iterator();
+						while (it2.hasNext()) {
+							String target = it2.next() + " " + type;
+							if (!target.startsWith("S:")
+									&& !target.equals(vertex)) {
+								addEdge(vertex, target, edge + " (" + counter++
+										+ ")");
+								System.out.println(vertex + " -> " + edge
+										+ " -> " + target);
+							}
+						}
+					}
+				}
+			}
+			sendEdgesToBack();
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	public List getWords(String line) {
+		List words = new ArrayList();
+		int index = line.indexOf("<a href=\"");
+		while (index > 0) {
+			index = line.indexOf("\">", index) + 2;
+			int end = line.indexOf("</a>", index);
+			String word = line.substring(index, end);
+			words.add(word);
+			index = line.indexOf("<a href=\"", end + 4);
+		}
+		return words;
+	}
+
+	public List getContent(String word) throws IOException {
+		URL url = new URL("http://wordnet.princeton.edu/perl/webwn");
+		URLConnection urlConn = url.openConnection();
+		urlConn.setDoInput(true);
+		urlConn.setDoOutput(true);
+		urlConn.setUseCaches(false);
+		urlConn.setRequestProperty("Content-Type",
+				"application/x-www-form-urlencoded");
+		DataOutputStream printout = new DataOutputStream(urlConn
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *printout*. The resource is referenced at lines 319, 323, 324. The resource is closed at line 325. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.flush throws an exception. Ensure *printout* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).